[mlir] Link libraries that aren't included in libMLIR to libMLIR

mgorny opened the pull request:
Use `mlir_target_link_libraries()` to link dependencies of libraries that are not included in libMLIR, to ensure that they link to the dylib when they are used in Flang.  Otherwise, they implicitly pull in all their static dependencies, effectively causing Flang binaries to simultaneously link to the dylib and to static libraries, which is never a good idea.

I have only covered the libraries that are used by Flang.  If you wish, I can extend this approach to all non-libMLIR libraries in MLIR, making MLIR itself also link to the dylib consistently.

mgorny committed:
[mlir] Link libraries that aren't included in libMLIR to libMLIR

Use `mlir_target_link_libraries()` to link dependencies of libraries
that are not included in libMLIR, to ensure that they link to the dylib
when they are used in Flang.  Otherwise, they implicitly pull in all
their static dependencies, effectively causing Flang binaries to
simultaneously link to the dylib and to static libraries, which is never
a good idea.

I have only covered the libraries that are used by Flang.  If you wish,
I can extend this approach to all non-libMLIR libraries in MLIR, making
MLIR itself also link to the dylib consistently.

nikic reviewed (commented):
This is a good idea, thanks.

> I have only covered the libraries that are used by Flang. If you wish, I can extend this approach to all non-libMLIR libraries in MLIR, making MLIR itself also link to the dylib consistently.

Yes, we should do this for everything.

mgorny committed:
Cover the remaining libraries

Hopefully I didn't mess anything up.  The tests build and pass for me
(except for these that failed already) both with MLIR_LINK_MLIR_DYLIB
and without.

mgorny commented:
Rebased and updated for all the libs.

nikic commented on mlir/test/lib/Analysis/CMakeLists.txt:
```diff
@@ -21,12 +21,18 @@ add_mlir_library(MLIRTestAnalysis
+
+# Since this library is excluded from libMLIR, link it to the dylib
+# to ensure that it can be used in flang without implicitly pulling in
+# static libraries.
```
Don't think we need this comment (or at least it's weird to have it in some places but not others).

mgorny replied:
Removed them and added a short note to `EXCLUDE_FROM_LIBMLIR` doc.

nikic reviewed (approved):
LGTM, thanks!

mgorny committed:
Replace comments with a note in EXCLUDE_FROM_LIBMLIR doc

mgorny commented:
Looks like I broke `BUILD_SHARED_LIBS=ON`. I'll revert and try to reproduce locally.
